### PR TITLE
chore: pass a custom logger from user for the builder factory

### LINKED ISSUE #538: use a custom logger from user for the builder

### PR DESCRIPTION
**Pull request by mojtaba-esk**
Closes #538 

local tests are passing and the debug logs of the builder is shown when the debug level is set:
```sh
=== RUN   TestRunSuite
time="2024-10-08T16:53:15+03:30" level=info msg="Current log level" file="log/logger.go:41" env_log_level=LOG_LEVEL log_level=debug
time="2024-10-08T16:53:15+03:30" level=debug msg="set state of instance" file="instance/state.go:41" instance=timeout-handler state=Preparing
time="2024-10-08T16:53:15+03:30" level=debug msg="no need to build and push image for instance" file="instance/build.go:164" image="docker.io/bitnami/kubectl:latest" instance=timeout-handler
time="2024-10-08T16:53:15+03:30" level=debug msg="set state of instance" file="instance/state.go:41" instance=timeout-handler state=Committed
time="2024-10-08T16:53:15+03:30" level=debug msg="the namespace will be deleted" file="knuu/knuu.go:140" namespace=20241008-165248-541
time="2024-10-08T16:53:16+03:30" level=debug msg="started statefulSet" file="instance/execution.go:337" instance=timeout-handler
time="2024-10-08T16:53:16+03:30" level=debug msg="set state of instance" file="instance/state.go:41" instance=timeout-handler state=Started
time="2024-10-08T16:53:23+03:30" level=debug msg="Service created successfully." file="traefik/traefik.go:259" service=traefik
time="2024-10-08T16:53:36+03:30" level=debug msg="proxy endpoint" file="knuu/knuu.go:237" endpoint="151.115.12.124:80"
    suite_setup_test.go:53: Scope: 20241008-165248-541
=== RUN   TestRunSuite/TestBuildFromGit
=== PAUSE TestRunSuite/TestBuildFromGit
=== RUN   TestRunSuite/TestBuildFromGitWithModifications
=== PAUSE TestRunSuite/TestBuildFromGitWithModifications
=== CONT  TestRunSuite/TestBuildFromGit
=== CONT  TestRunSuite/TestBuildFromGitWithModifications
    build_image_test.go:22: Creating new instance
    build_image_test.go:26: Building the image
    build_image_test.go:64: Creating new instance
    build_image_test.go:68: Setting git repo
time="2024-10-08T16:53:36+03:30" level=debug msg="set state of instance" file="instance/state.go:41" instance=build-from-git state=Preparing
time="2024-10-08T16:53:36+03:30" level=debug msg="Building image ttl.sh/bb7a5c161d9535a225a7fa67e665b4d8305204a82bff3c75cb925cf578f20b7e:24h from git repo github.com/celestiaorg/knuu" file="container/docker.go:143"
time="2024-10-08T16:53:36+03:30" level=debug msg="set state of instance" file="instance/state.go:41" instance=build-from-git-with-modifications state=Preparing
time="2024-10-08T16:53:36+03:30" level=debug msg="Building image ttl.sh/bb7a5c161d9535a225a7fa67e665b4d8305204a82bff3c75cb925cf578f20b7e:24h from git repo github.com/celestiaorg/knuu" file="container/docker.go:143"
time=2024-10-08T16:53:54+03:30 level=debug msg=build logs: Enumerating objects: 2995, done.
Counting objects: 100% (1676/1676), done.
Compressing objects: 100% (904/904), done.
Total 2995 (delta 1265), reused 895 (delta 764), pack-reused 1319 (from 1)
INFO[0002] Retrieving image manifest ubuntu:18.04       
INFO[0002] Retrieving image ubuntu:18.04 from registry index.docker.io 
INFO[0003] Retrieving image manifest ubuntu:18.04       
INFO[0003] Returning cached image manifest              
INFO[0003] Built cross stage deps: map[]                
INFO[0003] Retrieving image manifest ubuntu:18.04       
INFO[0003] Returning cached image manifest              
INFO[0003] Retrieving image manifest ubuntu:18.04       
INFO[0003] Returning cached image manifest              
INFO[0003] Executing 0 build triggers                   
INFO[0003] Building stage 'ubuntu:18.04' [idx: '0', base-idx: '-1'] 
INFO[0003] Checking for cached layer ttl.sh/bb7a5c161d9535a225a7fa67e665b4d8305204a82bff3c75cb925cf578f20b7e:24h:d9ec0812facda4186859133fdf2d9fdff77bd1feaf4bf75effde8c60df58c240... 
INFO[0003] No cached layer found for cmd RUN echo "${MESSAGE}" > /test.txt 
INFO[0003] Unpacking rootfs as cmd RUN echo "${MESSAGE}" > /test.txt requires it. 
INFO[0004] ARG MESSAGE="Hello, World!"                  
INFO[0004] No files changed in this command, skipping snapshotting. 
INFO[0004] RUN echo "${MESSAGE}" > /test.txt            
INFO[0004] Initializing snapshotter ...                 
INFO[0004] Taking snapshot of full filesystem...        
INFO[0005] Cmd: /bin/sh                                 
INFO[0005] Args: [-c echo "${MESSAGE}" > /test.txt]     
INFO[0005] Running: [/bin/sh -c echo "${MESSAGE}" > /test.txt] 
INFO[0005] Taking snapshot of full filesystem...        
INFO[0005] Pushing layer ttl.sh/bb7a5c161d9535a225a7fa67e665b4d8305204a82bff3c75cb925cf578f20b7e:24h:d9ec0812facda4186859133fdf2d9fdff77bd1feaf4bf75effde8c60df58c240 to cache now 
INFO[0005] CMD ["sleep", "infinity"]                    
INFO[0005] No files changed in this command, skipping snapshotting. 
WARN[0005] Error uploading layer to cache: getting tag for destination: repository can only contain the characters `abcdefghijklmnopqrstuvwxyz0123456789_-./`: bb7a5c161d9535a225a7fa67e665b4d8305204a82bff3c75cb925cf578f20b7e:24h 
INFO[0005] Pushing image to ttl.sh/bb7a5c161d9535a225a7fa67e665b4d8305204a82bff3c75cb925cf578f20b7e:24h 
INFO[0012] Pushed ttl.sh/bb7a5c161d9535a225a7fa67e665b4d8305204a82bff3c75cb925cf578f20b7e@sha256:decb5331e1940ad72de02fa4b0816ec13cb7f98f9d0ca905e6ff41412a509fa6 
 file=container/docker.go:156
    build_image_test.go:36: Image built
time="2024-10-08T16:53:54+03:30" level=debug msg="no need to build and push image for instance" file="instance/build.go:164" image="ttl.sh/bb7a5c161d9535a225a7fa67e665b4d8305204a82bff3c75cb925cf578f20b7e:24h" instance=build-from-git
time="2024-10-08T16:53:54+03:30" level=debug msg="build logs: Enumerating objects: 2995, done.\nCounting objects:   0% (1/1641)\rCounting objects:   1% (17/1641)\rCounting objects:   2% (33/1641)\rCounting objects:   3% (50/1641)\rCounting objects:   4% (66/1641)\rCounting objects:   5% (83/1641)\rCounting objects:   6% (99/1641)\rCounting objects:   7% (115/1641)\rCounting objects:   8% (132/1641)\rCounting objects:   9% (148/1641)\rCounting objects:  10% (165/1641)\rCounting objects:  11% (181/1641)\rCounting objects:  12% (197/1641)\rCounting objects:  13% (214/1641)\rCounting objects:  14% (230/1641)\rCounting objects:  15% (247/1641)\rCounting objects:  16% (263/1641)\rCounting objects:  17% (279/1641)\rCounting objects:  18% (296/1641)\rCounting objects:  19% (312/1641)\rCounting objects:  20% (329/1641)\rCounting objects:  21% (345/1641)\rCounting objects:  22% (362/1641)\rCounting objects:  23% (378/1641)\rCounting objects:  24% (394/1641)\rCounting objects:  25% (411/1641)\rCounting objects:  26% (427/1641)\rCounting objects:  27% (444/1641)\rCounting objects:  28% (460/1641)\rCounting objects:  29% (476/1641)\rCounting objects:  30% (493/1641)\rCounting objects:  31% (509/1641)\rCounting objects:  32% (526/1641)\rCounting objects:  33% (542/1641)\rCounting objects:  34% (558/1641)\rCounting objects:  35% (575/1641)\rCounting objects:  36% (591/1641)\rCounting objects:  37% (608/1641)\rCounting objects:  38% (624/1641)\rCounting objects:  39% (640/1641)\rCounting objects:  40% (657/1641)\rCounting objects:  41% (673/1641)\rCounting objects:  42% (690/1641)\rCounting objects:  43% (706/1641)\rCounting objects:  44% (723/1641)\rCounting objects:  45% (739/1641)\rCounting objects:  46% (755/1641)\rCounting objects:  47% (772/1641)\rCounting objects:  48% (788/1641)\rCounting objects:  49% (805/1641)\rCounting objects:  50% (821/1641)\rCounting objects:  51% (837/1641)\rCounting objects:  52% (854/1641)\rCounting objects:  53% (870/1641)\rCounting objects:  54% (887/1641)\rCounting objects:  55% (903/1641)\rCounting objects:  56% (919/1641)\rCounting objects:  57% (936/1641)\rCounting objects:  58% (952/1641)\rCounting objects:  59% (969/1641)\rCounting objects:  60% (985/1641)\rCounting objects:  61% (1002/1641)\rCounting objects:  62% (1018/1641)\rCounting objects:  63% (1034/1641)\rCounting objects:  64% (1051/1641)\rCounting objects:  65% (1067/1641)\rCounting objects:  66% (1084/1641)\rCounting objects:  67% (1100/1641)\rCounting objects:  68% (1116/1641)\rCounting objects:  69% (1133/1641)\rCounting objects:  70% (1149/1641)\rCounting objects:  71% (1166/1641)\rCounting objects:  72% (1182/1641)\rCounting objects:  73% (1198/1641)\rCounting objects:  74% (1215/1641)\rCounting objects:  75% (1231/1641)\rCounting objects:  76% (1248/1641)\rCounting objects:  77% (1264/1641)\rCounting objects:  78% (1280/1641)\rCounting objects:  79% (1297/1641)\rCounting objects:  80% (1313/1641)\rCounting objects:  81% (1330/1641)\rCounting objects:  82% (1346/1641)\rCounting objects:  83% (1363/1641)\rCounting objects:  84% (1379/1641)\rCounting objects:  85% (1395/1641)\rCounting objects:  86% (1412/1641)\rCounting objects:  87% (1428/1641)\rCounting objects:  88% (1445/1641)\rCounting objects:  89% (1461/1641)\rCounting objects:  90% (1477/1641)\rCounting objects:  91% (1494/1641)\rCounting objects:  92% (1510/1641)\rCounting objects:  93% (1527/1641)\rCounting objects:  94% (1543/1641)\rCounting objects:  95% (1559/1641)\rCounting objects:  96% (1576/1641)\rCounting objects:  97% (1592/1641)\rCounting objects:  98% (1609/1641)\rCounting objects:  99% (1625/1641)\rCounting objects: 100% (1641/1641)\rCounting objects: 100% (1641/1641), done.\nCompressing objects:   0% (1/900)\rCompressing objects:   1% (9/900)\rCompressing objects:   2% (18/900)\rCompressing objects:   3% (27/900)\rCompressing objects:   4% (36/900)\rCompressing objects:   5% (45/900)\rCompressing objects:   6% (54/900)\rCompressing objects:   7% (63/900)\rCompressing objects:   8% (72/900)\rCompressing objects:   9% (81/900)\rCompressing objects:  10% (90/900)\rCompressing objects:  11% (99/900)\rCompressing objects:  12% (108/900)\rCompressing objects:  13% (117/900)\rCompressing objects:  14% (126/900)\rCompressing objects:  15% (135/900)\rCompressing objects:  16% (144/900)\rCompressing objects:  17% (153/900)\rCompressing objects:  18% (162/900)\rCompressing objects:  19% (171/900)\rCompressing objects:  20% (180/900)\rCompressing objects:  21% (189/900)\rCompressing objects:  22% (198/900)\rCompressing objects:  23% (207/900)\rCompressing objects:  24% (216/900)\rCompressing objects:  25% (225/900)\rCompressing objects:  26% (234/900)\rCompressing objects:  27% (243/900)\rCompressing objects:  28% (252/900)\rCompressing objects:  29% (261/900)\rCompressing objects:  30% (270/900)\rCompressing objects:  31% (279/900)\rCompressing objects:  32% (288/900)\rCompressing objects:  33% (297/900)\rCompressing objects:  34% (306/900)\rCompressing objects:  35% (315/900)\rCompressing objects:  36% (324/900)\rCompressing objects:  37% (333/900)\rCompressing objects:  38% (342/900)\rCompressing objects:  39% (351/900)\rCompressing objects:  40% (360/900)\rCompressing objects:  41% (369/900)\rCompressing objects:  42% (378/900)\rCompressing objects:  43% (387/900)\rCompressing objects:  44% (396/900)\rCompressing objects:  45% (405/900)\rCompressing objects:  46% (414/900)\rCompressing objects:  47% (423/900)\rCompressing objects:  48% (432/900)\rCompressing objects:  49% (441/900)\rCompressing objects:  50% (450/900)\rCompressing objects:  51% (459/900)\rCompressing objects:  52% (468/900)\rCompressing objects:  53% (477/900)\rCompressing objects:  54% (486/900)\rCompressing objects:  55% (495/900)\rCompressing objects:  56% (504/900)\rCompressing objects:  57% (513/900)\rCompressing objects:  58% (522/900)\rCompressing objects:  59% (531/900)\rCompressing objects:  60% (540/900)\rCompressing objects:  61% (549/900)\rCompressing objects:  62% (558/900)\rCompressing objects:  63% (567/900)\rCompressing objects:  64% (576/900)\rCompressing objects:  65% (585/900)\rCompressing objects:  66% (594/900)\rCompressing objects:  67% (603/900)\rCompressing objects:  68% (612/900)\rCompressing objects:  69% (621/900)\rCompressing objects:  70% (630/900)\rCompressing objects:  71% (639/900)\rCompressing objects:  72% (648/900)\rCompressing objects:  73% (657/900)\rCompressing objects:  74% (666/900)\rCompressing objects:  75% (675/900)\rCompressing objects:  76% (684/900)\rCompressing objects:  77% (693/900)\rCompressing objects:  78% (702/900)\rCompressing objects:  79% (711/900)\rCompressing objects:  80% (720/900)\rCompressing objects:  81% (729/900)\rCompressing objects:  82% (738/900)\rCompressing objects:  83% (747/900)\rCompressing objects:  84% (756/900)\rCompressing objects:  85% (765/900)\rCompressing objects:  86% (774/900)\rCompressing objects:  87% (783/900)\rCompressing objects:  88% (792/900)\rCompressing objects:  89% (801/900)\rCompressing objects:  90% (810/900)\rCompressing objects:  91% (819/900)\rCompressing objects:  92% (828/900)\rCompressing objects:  93% (837/900)\rCompressing objects:  94% (846/900)\rCompressing objects:  95% (855/900)\rCompressing objects:  96% (864/900)\rCompressing objects:  97% (873/900)\rCompressing objects:  98% (882/900)\rCompressing objects:  99% (891/900)\rCompressing objects: 100% (900/900)\rCompressing objects: 100% (900/900), done.\nTotal 2995 (delta 1229), reused 864 (delta 733), pack-reused 1354 (from 1)\n\x1b[36mINFO\x1b[0m[0002] Retrieving image manifest ubuntu:18.04       \n\x1b[36mINFO\x1b[0m[0002] Retrieving image ubuntu:18.04 from registry index.docker.io \n\x1b[36mINFO\x1b[0m[0003] Retrieving image manifest ubuntu:18.04       \n\x1b[36mINFO\x1b[0m[0003] Returning cached image manifest              \n\x1b[36mINFO\x1b[0m[0003] Built cross stage deps: map[]                \n\x1b[36mINFO\x1b[0m[0003] Retrieving image manifest ubuntu:18.04       \n\x1b[36mINFO\x1b[0m[0003] Returning cached image manifest              \n\x1b[36mINFO\x1b[0m[0003] Retrieving image manifest ubuntu:18.04       \n\x1b[36mINFO\x1b[0m[0003] Returning cached image manifest              \n\x1b[36mINFO\x1b[0m[0003] Executing 0 build triggers                   \n\x1b[36mINFO\x1b[0m[0003] Building stage 'ubuntu:18.04' [idx: '0', base-idx: '-1'] \n\x1b[36mINFO\x1b[0m[0003] Checking for cached layer ttl.sh/bb7a5c161d9535a225a7fa67e665b4d8305204a82bff3c75cb925cf578f20b7e:24h:d9ec0812facda4186859133fdf2d9fdff77bd1feaf4bf75effde8c60df58c240... \n\x1b[36mINFO\x1b[0m[0003] No cached layer found for cmd RUN echo \"${MESSAGE}\" > /test.txt \n\x1b[36mINFO\x1b[0m[0003] Unpacking rootfs as cmd RUN echo \"${MESSAGE}\" > /test.txt requires it. \n\x1b[36mINFO\x1b[0m[0004] ARG MESSAGE=\"Hello, World!\"                  \n\x1b[36mINFO\x1b[0m[0004] No files changed in this command, skipping snapshotting. \n\x1b[36mINFO\x1b[0m[0004] RUN echo \"${MESSAGE}\" > /test.txt            \n\x1b[36mINFO\x1b[0m[0004] Initializing snapshotter ...                 \n\x1b[36mINFO\x1b[0m[0004] Taking snapshot of full filesystem...        \n\x1b[36mINFO\x1b[0m[0005] Cmd: /bin/sh                                 \n\x1b[36mINFO\x1b[0m[0005] Args: [-c echo \"${MESSAGE}\" > /test.txt]     \n\x1b[36mINFO\x1b[0m[0005] Running: [/bin/sh -c echo \"${MESSAGE}\" > /test.txt] \n\x1b[36mINFO\x1b[0m[0005] Taking snapshot of full filesystem...        \n\x1b[36mINFO\x1b[0m[0005] CMD [\"sleep\", \"infinity\"]                    \n\x1b[36mINFO\x1b[0m[0005] Pushing layer ttl.sh/bb7a5c161d9535a225a7fa67e665b4d8305204a82bff3c75cb925cf578f20b7e:24h:d9ec0812facda4186859133fdf2d9fdff77bd1feaf4bf75effde8c60df58c240 to cache now \n\x1b[36mINFO\x1b[0m[0005] No files changed in this command, skipping snapshotting. \n\x1b[33mWARN\x1b[0m[0005] Error uploading layer to cache: getting tag for destination: repository can only contain the characters `abcdefghijklmnopqrstuvwxyz0123456789_-./`: bb7a5c161d9535a225a7fa67e665b4d8305204a82bff3c75cb925cf578f20b7e:24h \n\x1b[36mINFO\x1b[0m[0005] Pushing image to ttl.sh/bb7a5c161d9535a225a7fa67e665b4d8305204a82bff3c75cb925cf578f20b7e:24h \n\x1b[36mINFO\x1b[0m[0013] Pushed ttl.sh/bb7a5c161d9535a225a7fa67e665b4d8305204a82bff3c75cb925cf578f20b7e@sha256:47b88f416c7038d8a7b39df32fa8d37f2fe94a4254fc28a2c6f6f980a76cee2d \n" file="container/docker.go:156"
time=2024-10-08T16:53:54+03:30 level=debug msg=set state of instance file=instance/state.go:41 instance=build-from-git state=Committed
    build_image_test.go:40: Starting instance
time=2024-10-08T16:53:54+03:30 level=debug msg=Generated image hash: 84a72f66837481462173d732f39d6b7705af88ead95b52d12b9dc8be93af9d01 file=container/docker.go:194
time=2024-10-08T16:53:54+03:30 level=debug msg=cannot use any cached image for instance file=instance/build.go:197 instance=build-from-git-with-modifications
time=2024-10-08T16:53:55+03:30 level=debug msg=started statefulSet file=instance/execution.go:337 instance=build-from-git
time=2024-10-08T16:53:55+03:30 level=debug msg=set state of instance file=instance/state.go:41 instance=build-from-git state=Started
    build_image_test.go:43: Instance started
    build_image_test.go:45: Getting file bytes
time=2024-10-08T16:54:14+03:30 level=debug msg=build logs: INFO[0001] Retrieving image manifest ttl.sh/bb7a5c161d9535a225a7fa67e665b4d8305204a82bff3c75cb925cf578f20b7e:24h 
INFO[0001] Retrieving image ttl.sh/bb7a5c161d9535a225a7fa67e665b4d8305204a82bff3c75cb925cf578f20b7e:24h from registry ttl.sh 
INFO[0002] Built cross stage deps: map[]                
INFO[0002] Retrieving image manifest ttl.sh/bb7a5c161d9535a225a7fa67e665b4d8305204a82bff3c75cb925cf578f20b7e:24h 
INFO[0002] Returning cached image manifest              
INFO[0002] Executing 0 build triggers                   
INFO[0002] Building stage 'ttl.sh/bb7a5c161d9535a225a7fa67e665b4d8305204a82bff3c75cb925cf578f20b7e:24h' [idx: '0', base-idx: '-1'] 
INFO[0002] Unpacking rootfs as cmd ADD --chown=root:root /home/hello.txt /home/hello.txt requires it. 
INFO[0004] Using files from context: [/kaniko/buildcontext/home/hello.txt] 
INFO[0004] ADD --chown=root:root /home/hello.txt /home/hello.txt 
INFO[0004] Taking snapshot of files...                  
INFO[0004] Pushing image to ttl.sh/84a72f66837481462173d732f39d6b7705af88ead95b52d12b9dc8be93af9d01:24h 
INFO[0009] Pushed ttl.sh/84a72f66837481462173d732f39d6b7705af88ead95b52d12b9dc8be93af9d01@sha256:95f7db225468fcce2b3753622b2752b7ad0228a2053dcb7392ba9a6e41c7fa13 
 file=container/docker.go:121
time=2024-10-08T16:54:14+03:30 level=debug msg=pushed new image for instance file=instance/build.go:208 image=ttl.sh/84a72f66837481462173d732f39d6b7705af88ead95b52d12b9dc8be93af9d01:24h instance=build-from-git-with-modifications
time=2024-10-08T16:54:14+03:30 level=debug msg=set state of instance file=instance/state.go:41 instance=build-from-git-with-modifications state=Committed
time=2024-10-08T16:54:15+03:30 level=debug msg=started statefulSet file=instance/execution.go:337 instance=build-from-git-with-modifications
time=2024-10-08T16:54:15+03:30 level=debug msg=set state of instance file=instance/state.go:41 instance=build-from-git-with-modifications state=Started
=== NAME  TestRunSuite
    suite.go:59: Cleaning up knuu...
    suite.go:63: Knuu is cleaned up
--- PASS: TestRunSuite (47.82s)
    --- PASS: TestRunSuite/TestBuildFromGit (26.41s)
    --- PASS: TestRunSuite/TestBuildFromGitWithModifications (46.72s)
PASS
ok  	github.com/celestiaorg/knuu/e2e/system	94.574s
```

## Summary by CodeRabbit

- **New Features**
	- Enhanced logging capabilities in the BuilderFactory structure.
	- Introduced a new structured format for BuilderFactory options.

- **Bug Fixes**
	- Added specific error messages for empty fields related to image and build context management.

- **Refactor**
	- Updated methods to utilize the new BuilderFactoryOptions struct for improved parameter handling.
	- Refactored the instantiation of the BuilderFactory in the SetImage and SetGitRepo methods.

### CODE REVIEW

File: pkg/container/docker.go
Comment (MSevey) [inline pkg/container/docker.go:208]: Is image name really required? Feels like we could just auto generate one on the fly.

File: pkg/container/docker.go
Comment (MSevey) [inline pkg/container/docker.go:207]: Shouldn't image builder be validated here? Since we check for it explicitly in multiple places?

Adding the check here should make it impossible to create the BuilderFactory without it, so we could clean up all those inline checks.

File: pkg/container/docker.go
Comment (MSevey) [inline pkg/container/docker.go:214]: Same here, if a logger isn't passed in, can't we just create one?

File: pkg/container/docker.go
Comment (MSevey) [reply pkg/container/docker.go:126]: We also do this multiple times. Might as well create a helper for when we want to print and explicitly disable quotes for that log line.

File: pkg/container/docker.go
Comment (mojtaba-esk) [reply pkg/container/docker.go:207]: that could be neater, however the image builder is not a requirement as the user can set an image that is already built. Image builder is required only if we want to build from git or from a local directory.
Do you think we should make it a requirement?

File: pkg/container/docker.go
Comment (mojtaba-esk) [reply pkg/container/docker.go:208]: yes we need to have it and the generated image (i.e. built image)  will use that name. it is essential for the long running tests to have static names for all resources including images.

File: pkg/container/docker.go
Comment (mojtaba-esk) [reply pkg/container/docker.go:214]: The logger is required throughout the project so we use a unique logger object to handle all the logs the same way. i.e. it is more consistent to have only one logger object and pass it throughout the entire code base.

Comment (MSevey) [reply pkg/container/docker.go:204]: fwiw making it required here isn't the same as ensuring it is the same across the code base. 
but I won't block on this. You can resolve this.

File: pkg/container/docker.go
Comment (MSevey) [reply pkg/container/docker.go:197]: To preface, If this is a bigger change we resolve into a follow up issue to unblock this PR. 

It sounds like there are 2 distinct user flows. 
1. No image exists yet, we need to build it
2. An image exists, and we need to use it.

So it would seem that the more concrete thing that needs to be check is the built image itself. In absence of that, it seems we are checking the name or the image builder as proxies. 

Is there a way to know if the image was already built or it needs to be built? That might clean up the logic here to make it more specific to each scenario vs having some of the context hidden.

Either way, like i said to start, if this is a deeper rabbit hole we can discuss this more in an issue or offline and unblock this PR.

**Review by MSevey — APPROVED** [review]

File: pkg/container/docker.go
Comment (mojtaba-esk) [reply pkg/container/docker.go:197]: I think it is covered in the #525 as it requires a refactor around the image names and the main focus of this PR is to introduce logger to the builderFactory. Therefore, I agree with the option that we unlock this PR and track it in the other issue.

File: pkg/container/docker.go
Comment (mojtaba-esk) [reply pkg/container/docker.go:204]: an addition point is the fact that this BuilderFactory is used internally by the knuu and the external user is not exposed to it. So, we have more flexibility to refactor it the way we think is optimized in future.

**Pull request merged**
